Issue 495 - jansson style rpc calls

chu11 opened the pull request:
Various refactoring to update rpc calls from json-c to jansson style.  Such as calling ```flux_rpcf()``` over ```flux_rpc()```.

There are more that can be done, but I limited this PR to only those rpcs that dealt with simple int, string, etc. types.  Dealing with more complex objects/arrays require using the jansson library to create/manipulate said objects/arrays instead of the json-c library, and that's refactoring/cleanup for another day.

Most fixes should be self explanatory hopefully.  The noteable exception is https://github.com/flux-framework/flux-core/pull/1002/commits/37be05995855e21573b983422f58fd875655665d.  The jansson style functions are pickier on typing.  So a ```"1"``` (in quotes) is a string and not an integer, so I updated ```flux-cron``` to send the right type.

chu11 commented:
Oh, there is one other thing that may be viewed as quirky.  The ```flux_rmmod_json_decode()``` and ```flux_rmmod_json_encode()``` functions only encode/decode a single ```name``` string, so I refactored them out.  Unsure if they would be desired to be left in for some general API abstraction desires.

I didn't remove the functions outright b/c ```flux-sched``` uses ```flux_rmmod_json_decode()``` for the moment.

garlick commented:
One suggestion is to remove the `#include shortjson.h` or `#include json.h`  wherever possible.

As noted in  0de293f81e8f80f50b76173c7ead20a0ff1d3c6e, the internal json-c library is "re-namespaced" using the preprocessor to put a json_c_ prefix on all functions.  The redefinition covers the source files used to build libjson-c.a, and any Flux C modules that include the above files.

There might not be a problem as long as overlapping names are avoided, as is probably the case with all your changes.  However, just to be safe since json code could be introduced in the future and the preferred library is jansson, if you've weaned C module off of json-c, it is best to drop the include.

chu11 commented:
@garlick Ahh, I had missed that.  I think I even found a few places we could remove ```jansson.h``` as well.

While going through cleanup of ```shortjson.h```, I realized that I could also cleanup usage of ```flux_event_encode()```.  That is one I had missed earlier.  I'll try to add more cleanup with that as well.

chu11 commented:
Reworked the PR.  Added some additional cleanup using ```flux_event_encodef()``` and ```flux_event_decodef()```.  Removed ```shortjson.h``` where possible.  

Also elected to not refactor out use of ```flux_rmmod_json_decode()``` and ```flux_rmmod_json_encode()```.  I think it's better to refactor those in another way.

chu11 commented:
Looking through code coverage, unsurprisingly lots of error paths missed.

It appears no test coverage of ssh connector exists b/c it missed everything there.  Lots of live module misses, I suspect from prior removal of topo & up in https://github.com/flux-framework/flux-core/pull/960

grondo commented:
> Looking through code coverage, unsurprisingly lots of error paths missed.

Yeah, usually it is difficult to test error paths. As long as the project coverage doesn't drop appreciably, I take the code coverage checks as advisory.

> It appears no test coverage of ssh connector exists b/c it missed everything there

Hm, that is concerning ... on master there is good coverage of ssh connector

https://codecov.io/gh/flux-framework/flux-core/src/master/src/connectors/ssh/ssh.c

garlick commented:
Note in #638 the suggestion that we remove the module.  We need to start from scratch on that one.  Want to just remove it here instead of refactoring?

(Sorry I didn't just do it earlier and save you the extra work!)

chu11 commented:
@grondo Ahh, I spoke too soon.  It appears ```op_event_subscribe``` and ```op_event_unsubscribe``` are not covered in the ssh connector, which are the only two places I cleaned up in the ssh connector.  Let me see if some tests can be done to done to correct that.

chu11 commented:
@garlick Ahh, wasn't aware of that issue.  I'll just remove it in another PR and rebase here.

grondo commented:
@chu11, cool good find!

chu11 commented:
Just sent PR #1003 and #1004 to address the poor coverage on the live module and ssh connector.  Once those are in, can rebase and hopefully the coverage will be a lot better here.

grondo commented:
Looks good, nice cleanup!

garlick commented:
Well done!
